src/System.Threading.Tasks/tests/System.Runtime.CompilerServices/AsyncTaskMethodBuilderTests.cs

```diff
@@ -233,244 +227,232 @@ public static void AsyncMethodBuilderCreate_SetExceptionTest2()
-                }
-                Assert.Throws<InvalidOperationException>(
-                   () => { atmb.SetResult(44); });
-                Assert.Throws<InvalidOperationException>(
-                   () => { atmb.SetException(new Exception()); });
-            }
+                t.GetAwaiter().GetResult();
+            });
+            Assert.Same(oce, e);
+            Assert.Throws<InvalidOperationException>(() => { atmb.SetResult(44); });
+            Assert.Throws<InvalidOperationException>(() => { atmb.SetException(new Exception()); });
         }
 
-        // random other stuff, e.g. exception dispatch info
         [Fact]
-        public static void RunAsyncAdditionalBehaviorsTests()
+        public static void TaskMethodBuilder_TaskIsCached()
         {
-            {
-                var atmb = new AsyncTaskMethodBuilder();
-                var t1 = atmb.Task;
-                var t2 = atmb.Task;
-                Assert.True(t1 != null, "     > FAILURE. Task should have been initialized.");
-                Assert.True(t2 != null, "     > FAILURE. Task should have been initialized.");
-                Assert.True(t1 == t2, "     > FAILURE. Task should be cached once initialized.");
-            }
-            {
-                var atmb = new AsyncTaskMethodBuilder<int>();
-                var t1 = atmb.Task;
-                var t2 = atmb.Task;
-                Assert.True(t1 != null, "     > FAILURE. Task should have been initialized.");
-                Assert.True(t2 != null, "     > FAILURE. Task should have been initialized.");
-                Assert.True(t1 == t2, "     > FAILURE. Task should be cached once initialized.");
-            }
+            var atmb = new AsyncTaskMethodBuilder();
+            var t1 = atmb.Task;
+            var t2 = atmb.Task;
+            Assert.NotNull(t1);
+            Assert.NotNull(t2);
+            Assert.Same(t1, t2);
         }
 
         [Fact]
-        public static void RunAsyncAdditionalBehaviorsTests_NegativeCases()
+        public static void TaskMethodBuilderT_TaskIsCached()
         {
-            // Test ExceptionDispatchInfo usage
-            //if (Thread.CurrentThread.CurrentCulture.ThreeLetterWindowsLanguageName == "ENU") // only on ENU because of string comparisons
-            //{
-            {
-                var tcs = new TaskCompletionSource<int>();
-                try { throw new InvalidOperationException(); }
-                catch (Exception e) { tcs.SetException(e); }
-                Assert.True(ValidateFaultedTask(tcs.Task), "     > FAILURE. Task's stack trace is incorrect");
-            }
+            var atmb = new AsyncTaskMethodBuilder<int>();
+            var t1 = atmb.Task;
+            var t2 = atmb.Task;
+            Assert.NotNull(t1);
+            Assert.NotNull(t2);
+            Assert.Same(t1, t2);
+        }
```

Comment:
It'd be nice to add a variation to this test that validates an important optimization these have of using a cached task if the method builder is completed synchronously and there's a cached task available, e.g.

```C#
[Fact]
public static void TaskMethodBuilder_UsesCompletedCache()
{
    var atmb1 = new AsyncTaskMethodBuilder();
    var atmb2 = new AsyncTaskMethodBuilder();
    atmb1.SetResult();
    atmb2.SetResult();
    Assert.Same(atmb1.Task, atmb2.Task);
}

[Theory]
[InlineData(true)]
[InlineData(false)]
public static void TaskMethodBuilderBoolean_UsesCompletedCache(bool result)
{
    TaskMethodBuilderT_UsesCompletedCache(result, shouldBeCached);
}

[Theory]
[InlineData(0, true)]
[InlineData(5, true)]
[InlineData(-5, true)]
[InlineData(42, false)]
public static void TaskMethodBuilderInt32_UsesCompletedCache(int result, bool shouldBeCached)
{
    TaskMethodBuilderT_UsesCompletedCache(result, shouldBeCached);
}

[Theory]
[InlineData((string)null, true)]
[InlineData("test", false)]
public static void TaskMethodBuilderRef_UsesCompletedCache(string result, bool shouldBeCached)
{
    TaskMethodBuilderT_UsesCompletedCache(result, shouldBeCached);
}

private static void TaskMethodBuilderT_UsesCompletedCache<T>(T result, bool shouldBeCached)
{
    var atmb1 = new AsyncTaskMethodBuilder<T>();
    var atmb2 = new AsyncTaskMethodBuilder<T>();

    atmb1.SetResult(result);
    atmb2.SetResult(result);

    Assert.Equal(shouldBeCached, object.ReferenceEquals(atmb1.Task, atmb2.Task));
}
```